lnwallet/channel.go

```diff
@@ -4622,6 +4622,17 @@ func (lc *LightningChannel) SignNextCommitment() (*NewCommitState, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	// We'll need to send over the signatures to the remote party in the
```

Comment:
Took this commit over into part3, added this commit message:

```
To make sure we attempt to read the results of the sig batches in the
same order they're processed, we sort them _before_ submitting them to
the batch processor.
Otherwise it might happen that we try to read on a result channel that
was never sent on because we aborted due to an error.
We also use slices.SortFunc now which doesn't use reflection and might
be slightly faster.
```

Reply:
Awesome!

I'm wondering now how we should handle these multiple `lnd` branches, if we want to cut a release with this included, before `0-19-staging` gets fully reviewed and pulled into `master`.

Reply:
Basically we're only keeping the `0-19-staging` in order for us to be able to cut a release before the rebased versions land in `master`. So as long as we don't have everything in `master`, we'll keep adding bugfixes here and then reference this branch in `litd`, using the `experimental` suffix.
Once everything landed in `master`, we'll cut `lnd v0.18.4-beta` and a non-experimental `litd` release.

Reply:
Ah ok, so the key insight here seems to be: 
> Otherwise it might happen that we try to read on a result channel that was never sent on because we aborted due to an error.

Reply:
Though I still don't see why this change is needed at all. 

Note that in master, we don't sort anything here: https://github.com/lightningnetwork/lnd/blob/750770e1904dd7db54b2cf1aa510d5e92308e290/lnwallet/channel.go#L3867-L3874

This seems to have been added only in the side branch. 

We don't need to sort anything here as things are already in the correct order, as we iterate over a slice which is ordered according to the htlc ID: https://github.com/lightningnetwork/lnd/blob/750770e1904dd7db54b2cf1aa510d5e92308e290/lnwallet/channel.go#L3105-L3111

When we go to validate, we iterate over the `TxOut`, then use an index to locate the HTLC that maps to a given output index: https://github.com/lightningnetwork/lnd/blob/750770e1904dd7db54b2cf1aa510d5e92308e290/lnwallet/channel.go#L4523-L4528

So unless we have a failing unit test to show this actually fixes something, I think we should drop it from the diff.

Reply:
Ah no, I'm wrong re the above, we sort here, right before we read the response: https://github.com/lightningnetwork/lnd/blob/750770e1904dd7db54b2cf1aa510d5e92308e290/lnwallet/channel.go#L3913-L3918

Reply:
With that said, I think if we're going to make this change, then we need a test to demonstrate the concrete impact. Also based on the commit message at the top of this comment thread, it seems the correct solution would be to thread through either a context or use the quit channel.

Reply:
> With that said, I think if we're going to make this change, then we need a test to demonstrate the concrete impact.

Fair enough; seems like the `TestMaxAcceptedHTLCs` test could be forked to add a fallible aux signer that causes the issue I'm describing. Working on that now.

> Also based on the commit message at the top of this comment thread, it seems the correct solution would be to thread through either a context or use the quit channel.

I think that would be a helpful change, though still separate from the sorting issue I've described. AFAICT the correct quit channel to propagate would be `channelLink.quit`:

https://github.com/lightningnetwork/lnd/blob/fb66bd2e96c3435ad2c7c49fa73fea6d6967a552/htlcswitch/link.go#L386

Which is the only non-test caller of `SignNextCommitment`.

Reply:
A test to demonstrate the issue and validate the fix is now in the second commit. To test:

```bash
git checkout jharveyb/aux_signer_batching_fixes
git checkout HEAD~3

# This test run should time out, and show that SignNextCommitment was stuck on reading the sig job response
make unit pkg=lnwallet case=TestAuxSignerShutdown timeout=10s

git checkout jharveyb/aux_signer_batching_fixes

# This should pass
make unit pkg=lnwallet case=TestAuxSignerShutdown timeout=10s
```

Reply:
I also added the propagation of the quit channel in here as the last commit, though I'm not sure that's the best way to actually add that parameter.

I did test in the playground that passing a nil channel in the tests is safe, and that arg is not nil outside of tests. It seems like setting that quit channel when constructing a `LightningChannel` would be better, but also a more invasive change. Open to doing that a different way.